### PR TITLE
Low Acceleration Mode At Certain Elevator Height

### PR DESCRIPTION
**Pull request by johnli01**
Code may be inefficiently written

### CODE REVIEW

File: Competition2018/src/main/java/competition/subsystems/drive/DriveSubsystem.java [hunk 1]
```diff
@@ -51,6 +51,8 @@
     private double leftAccum;
     private double rightAccum;
     
+    private boolean toggleAccelerationLimit;
```
Comment (WasabiFan) [inline Competition2018/src/main/java/competition/subsystems/drive/DriveSubsystem.java]: Can this be removed?

File: Competition2018/src/main/java/competition/subsystems/drive/commands/ArcadeDriveWithJoysticksCommand.java [hunk 2]
```diff
@@ -14,14 +15,33 @@
...
     final OperatorInterface oi;
     final HeadingAssistModule ham;
+    
+    private double translateAdjustedPower;
+    private double rotateAdjustedPower;
+    private double turn;
+    
+    private double nextTranslateInput;
+    private double currentTranslateOutput;
+    private double translateInputDelta;
```
Comment (WasabiFan) [inline Competition2018/src/main/java/competition/subsystems/drive/commands/ArcadeDriveWithJoysticksCommand.java]: Can this be a local variable? (i.e., does it actually need to be saved?)

File: Competition2018/src/main/java/competition/subsystems/drive/commands/ArcadeDriveWithJoysticksCommand.java [hunk 3]
```diff
@@ -14,14 +15,33 @@
 public class ArcadeDriveWithJoysticksCommand extends BaseCommand {
 
     final DriveSubsystem driveSubsystem;
+    final ElevatorSubsystem elevatorSubsystem;
     final OperatorInterface oi;
     final HeadingAssistModule ham;
+    
+    private double translateAdjustedPower;
+    private double rotateAdjustedPower;
+    private double turn;
```
Comment (WasabiFan) [inline Competition2018/src/main/java/competition/subsystems/drive/commands/ArcadeDriveWithJoysticksCommand.java]: What's this variable used for?

File: Competition2018/src/main/java/competition/subsystems/drive/commands/ArcadeDriveWithJoysticksCommand.java [hunk 4]
```diff
@@ -14,14 +15,33 @@
...
+    
+    private double nextTranslateInput;
+    private double currentTranslateOutput;
+    private double translateInputDelta;
+    
+    private double nextRotateInput;
```
Comment (WasabiFan) [inline Competition2018/src/main/java/competition/subsystems/drive/commands/ArcadeDriveWithJoysticksCommand.java]: Do these need to be public? Or can they be kept local?

File: Competition2018/src/main/java/competition/subsystems/drive/commands/ArcadeDriveWithJoysticksCommand.java [hunk 5]
```diff
@@ -30,17 +50,43 @@ public ArcadeDriveWithJoysticksCommand(OperatorInterface oi, DriveSubsystem driv
...
-        double turn = ham.calculateHeadingPower(rotate);
-        driveSubsystem.drive(new XYPair(0, translate), turn);
+        nextTranslateInput = oi.driverGamepad.getLeftVector().y;
+        nextRotateInput = MathUtils.squareAndRetainSign(oi.driverGamepad.getRightVector().x);
+        
+        translateInputDelta = nextTranslateInput - currentTranslateOutput;
+        rotateInputDelta = nextRotateInput - currentRotateOutput;
+        
+        // Rotate
+        if (rotateInputDelta > 0.2 && elevatorSubsystem.getCurrentHeightInInches() > 40) {
```
Comment (WasabiFan) [inline Competition2018/src/main/java/competition/subsystems/drive/commands/ArcadeDriveWithJoysticksCommand.java]: Let's make all these constants `DoubleProperty`s so we can change them in competition.

File: Competition2018/src/main/java/competition/subsystems/drive/commands/ArcadeDriveWithJoysticksCommand.java [hunk 6]
```diff
@@ -30,17 +50,43 @@ public ArcadeDriveWithJoysticksCommand(OperatorInterface oi, DriveSubsystem driv
...
+            rotateAdjustedPower += 0.2;
+            turn = ham.calculateHeadingPower(rotateAdjustedPower);
```
Comment (WasabiFan) [inline Competition2018/src/main/java/competition/subsystems/drive/commands/ArcadeDriveWithJoysticksCommand.java]: Can we set the adjusted power in each if statement branch and then have one line at the end which actually runs the final output calculations and sets the powers? i.e., decide what you want to do, and then do it.

File: Competition2018/src/main/java/competition/subsystems/drive/commands/ArcadeDriveWithJoysticksCommand.java (hunk 5 above)
Comment (WasabiFan) [inline Competition2018/src/main/java/competition/subsystems/drive/commands/ArcadeDriveWithJoysticksCommand.java]: Wrap the whole thing in a check for the elevator height -- we don't need to run any of this if the elevator is below the height limit.

File: Competition2018/src/main/java/competition/subsystems/drive/commands/ArcadeDriveWithJoysticksCommand.java [hunk 7]
```diff
@@ -30,17 +50,43 @@ public ArcadeDriveWithJoysticksCommand(OperatorInterface oi, DriveSubsystem driv
...
-        double translate = oi.driverGamepad.getLeftVector().y;
-        double rotate = MathUtils.squareAndRetainSign(oi.driverGamepad.getRightVector().x);
-                
-        double turn = ham.calculateHeadingPower(rotate);
-        driveSubsystem.drive(new XYPair(0, translate), turn);
+        nextTranslateInput = oi.driverGamepad.getLeftVector().y;
+        nextRotateInput = MathUtils.squareAndRetainSign(oi.driverGamepad.getRightVector().x);
+        
+        translateInputDelta = nextTranslateInput - currentTranslateOutput;
+        rotateInputDelta = nextRotateInput - currentRotateOutput;
```
Comment (WasabiFan) [inline Competition2018/src/main/java/competition/subsystems/drive/commands/ArcadeDriveWithJoysticksCommand.java]: Do we ever set `currentRotateOutput`?

**Review by WasabiFan — CHANGES_REQUESTED** [review]
The logic itself looks good over all; these are some clean-up and maintainability suggestions.

File: Competition2018/src/main/java/competition/subsystems/drive/commands/ArcadeDriveWithJoysticksCommand.java (hunk 7 above)
Comment (johnli01) [reply Competition2018/src/main/java/competition/subsystems/drive/commands/ArcadeDriveWithJoysticksCommand.java]: Yeah I realized that fat error, fixed it luckily

File: Competition2018/src/main/java/competition/subsystems/drive/commands/ArcadeDriveWithJoysticksCommand.java (hunk 4 above)
Comment (johnli01) [reply Competition2018/src/main/java/competition/subsystems/drive/commands/ArcadeDriveWithJoysticksCommand.java]: Most likely local

File: Competition2018/src/main/java/competition/subsystems/drive/DriveSubsystem.java (hunk 1 above)
Comment (johnli01) [reply Competition2018/src/main/java/competition/subsystems/drive/DriveSubsystem.java]: Yessir

File: Competition2018/src/main/java/competition/subsystems/drive/commands/ArcadeDriveWithJoysticksCommand.java (hunk 3 above)
Comment (johnli01) [reply Competition2018/src/main/java/competition/subsystems/drive/commands/ArcadeDriveWithJoysticksCommand.java]: Helped me keep track of things, these could be changed/removed

File: Competition2018/src/main/java/competition/subsystems/drive/commands/ArcadeDriveWithJoysticksCommand.java (hunk 2 above)
Comment (johnli01) [reply Competition2018/src/main/java/competition/subsystems/drive/commands/ArcadeDriveWithJoysticksCommand.java]: Yeah they can be a local variable, i can make the switch

File: Competition2018/src/main/java/competition/subsystems/drive/commands/ArcadeDriveWithJoysticksCommand.java (hunk 5 above)
Comment (johnli01) [reply Competition2018/src/main/java/competition/subsystems/drive/commands/ArcadeDriveWithJoysticksCommand.java]: Got it

Comment (johnli01) [reply Competition2018/src/main/java/competition/subsystems/drive/commands/ArcadeDriveWithJoysticksCommand.java]: yes chef

File: Competition2018/src/main/java/competition/subsystems/drive/commands/ArcadeDriveWithJoysticksCommand.java (hunk 6 above)
Comment (johnli01) [reply Competition2018/src/main/java/competition/subsystems/drive/commands/ArcadeDriveWithJoysticksCommand.java]: So basically create the variable inside the if statement?

**Pull request merged**